### PR TITLE
Sanitize column only when applying snapshot diff

### LINKED ISSUE #18247: Can't edit properties of M2O fields after creation
### Describe the Bug

In an created M2O field; when editing the field properties again it fails with 

`[INTERNAL_SERVER_ERROR] alter table `assembly` modify `installation` varchar(36) not null - ER_FK_COLUMN_CANNOT_CHANGE: Cannot change column 'installation': used in a foreign key constraint 'assembly_installation_foreign'`

In my case I have an "assembly" collection with a M2O field with a relation the "installation" collection.

The ID is a UUID 

### To Reproduce

add a M2O field relation.
try to edit anything (ex. null behaviour ) on the properties (after creation)

10.6.12-MariaDB MariaDB Server

### Directus Version

v9.25.1

### Hosting Strategy

Self-Hosted (Custom)

### LINKED ISSUE #18299: Cannot add translations to the primary key of a collection
### Describe the Bug

It fails with the following error when adding a translation entry to the primary key (or even any change to the primary key)

{
  "errors": [
    {
      "message": "alter table \"t_parts\" alter column \"id\" drop not null - field \"id\" is a primary key",
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR"
     
    }
  ]
}

It seems that the constraint 'not null' is always dropped when alter a field that is defined as not null.

### To Reproduce

Try to alter something for a primary key and save.

### Directus Version

9.5.2

### Hosting Strategy

Self-Hosted (Docker Image)

### LINKED ISSUE #18379: Incorrect schema difference being detected
### Describe the Bug

In 9.26, when modifying metadata on a field Directus is detecting a schema change when none has occurred - for example, adding a value to a dropdown list.

The issue is caused by line 402 in fields.ts: 

    if (!isEqual(sanitizeColumn(existingColumn), hookAdjustedField.schema)) {

sanitizeColumn has a picklist that excludes these fields:
 - comment
 - foreign_key_schema
 - schema

hookAdjustedField.schema contains these fields, while sanitizeColumn(existingColumn) does not, which causes Directus to erroneously detect a schema change when none has happened.

To fix the issue, either both fields need to be wrapped by sanitizeColumn() or neither:

    if (!isEqual(sanitizeColumn(existingColumn), sanitizeColumn(hookAdjustedField.schema as Column)))

Unfortunately not able to create a PR for the fix for this due to BSL, but wanted to share the issue and fix to (hopefully) save the Directus team some time. 

### To Reproduce

Modify any field in Directus, metadata only. Directus will issue an ALTER TABLE command even though nothing has changed.

This impacts any customer that may have a view bound to the database table, the ALTER TABLE command will fail due to the bound view.

### Directus Version

v9.26

### Hosting Strategy

Self-Hosted (Custom)

### LINKED ISSUE #18472: Directus attempts to modify data type of primary key when field is set to readonly
### Describe the Bug

I have an pre-existing table `announcement` with this DDL:

```
CREATE TABLE `announcement` (
  `id` int NOT NULL AUTO_INCREMENT,
  `title` varchar(255) DEFAULT NULL,
  `description` text,
  `start_date` datetime DEFAULT NULL,
  `end_date` datetime DEFAULT NULL,
  `qr_html` longtext,
  `qr_url` mediumtext,
  `link` mediumtext,
  PRIMARY KEY (`id`),
) ENGINE=InnoDB AUTO_INCREMENT=300 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
```

When setting the field `id` to readonly via the data studio, directus attempts to modify the data type of the column resulting in the error:
```
alter table `announcement` modify `id` int unsigned not null auto_increment primary key - ER_MULTIPLE_PRI_KEY: Multiple primary key defined
```

My expectation is that directus shouldn't be modifying the `id` column and should only be flagging the `readonly` column for the `id` field in the directus_fields table. The fact that I can manually alter that `readonly` column and have it work as expected tells me that modifying the actual `id` column is unnecessary.

### To Reproduce

Create the table from above and set the id to readonly in the data studio.

### Directus Version

10.0.0

### Hosting Strategy

Self-Hosted (Custom)

### PR DESCRIPTION
**Pull request by licitdev**
`sanitizeColumn` was introduced in #18048 to remove vendor specific properties when applying diffs. 
It resulted the incorrect "dirty" identification of schema when updating via `/fields/{collection}`.  

> **Warning**
> - [x] Revert testing for all vendors before merging

Fixes #18247
Fixes #18472
Fixes #18379
Fixes #18299
Closes ENG-902

### CODE REVIEW

**Review by paescuj — APPROVED** [review]
LGTM 👍 

Fixes #18472, #18379, #18299 too 🚀

File: api/src/services/fields.ts [hunk 1]
```diff
@@ -399,7 +399,10 @@ export class FieldsService {
 			if (hookAdjustedField.schema) {
 				const existingColumn = await this.schemaInspector.columnInfo(collection, hookAdjustedField.field);
 
-				if (!isEqual(sanitizeColumn(existingColumn), hookAdjustedField.schema)) {
+				// Sanitize column only when applying snapshot diff
+				const columnToCompare = opts?.bypassLimits ? sanitizeColumn(existingColumn) : existingColumn;
```
Comment (paescuj) [inline api/src/services/fields.ts:403]: Only a small question: How safe is it to bet on this option? I think it's okay, but I wonder if we should pass another option à la `isSnapshotChange` when applying snapshot diffs to be able to recognize for sure 🤔

**Review by paescuj — APPROVED** [review]

File: api/src/services/fields.ts (hunk 1 above)
Comment (paescuj) [reply api/src/services/fields.ts]: As discussed on Discord we can keep it like this 👍

**Review by azrikahar — APPROVED** [review]
Nice catch! LGTM 👍

**Comment by yottoo** [comment]
Could there be a new minor release for these bug fixes?

**Comment by iksent** [comment]
Am I experiencing another issue?

https://github.com/directus/directus/pull/14128#issuecomment-1552746900

I tried to update to the latest `10.1.1` version and can't run snapshot applying anyway :(

**Comment by yottoo** [comment]
> Am I experiencing another issue?
> 
> [#14128 (comment)](https://github.com/directus/directus/pull/14128#issuecomment-1552746900)
> 
> I tried to update to the latest `10.1.1` version and can't run snapshot applying anyway :(

make sure that direcuts has been updated to 10.0.0

don't just `npm update`
but rather:
```
npm install directus@10
npx directus database migrate:latest
```

**Comment by iksent** [comment]
@yottoo, The problem was in different databases. I switched to MariaDB locally, made a new snapshot and successfully imported it on remote MariaDB.

Sorry for confusing.

**Pull request merged**